### PR TITLE
[Feature]: adds support for Gumbel Distribution

### PR DESCRIPTION
**Pull request by soumyasen1809**
Solves #189 

Added support for Gumbel distribution

### CODE REVIEW

**Comment by YeungOnion** [comment]
Looks good! Could you mention where you got the test values from in this PR?

**Comment by soumyasen1809** [comment]
> Looks good! Could you mention where you got the test values from in this PR?

Hi, I got the values from Wolfram alpha https://www.wolframalpha.com/input/?i=gumbel+distribution
Basically, I used the formula in the `Math Input` section and added the input values to get the results.

**Comment by YeungOnion** [comment]
Good reference. "As accurate as Wolfram" is absolutely acceptable.

Would you be able to reference some test suites in other libraries to find tail behavior we should test? My concern is because it is doubly exponential.

**Comment by YeungOnion** [comment]
additionally, two admin-esque things
1. Would you write up a conventional commit as a `feat`? (You can squash or not) But this will help in generating a changelog
2. Would you make this PR up to but not including the merge? Linear history is nice where we can have it.

**Comment by soumyasen1809** [comment]
Hi @YeungOnion , some unclear points:

> 1. Would you write up a conventional commit as a `feat`? (You can squash or not) But this will help in generating a changelog

I am not sure what you mean by "write up a conventional commit as a `feat`". Do you mean the title of the PR? It already says - [Feature]: adds support for Gumbel Distribution.
Or do you mean I should git commit --amend the last commit and add the [Feature] tag?
I don't mind keeping the commits unsquashed.

> 2. Would you make this PR up to but not including the merge? Linear history is nice where we can have it.
Do you mean to do a rebase to the current commit? I just did it.

**Comment by soumyasen1809** [comment]
> Would you be able to reference some test suites in other libraries to find tail behavior we should test? My concern is because it is doubly exponential.

I am not sure of any other Rust statistics lib that does Gumbel distribution. Sorry :(

**Comment by YeungOnion** [comment]
Oh I can clear those up

> Or do you mean I should git commit --amend the last commit and add the [Feature] tag?

Yes, if just the first of these commits could instead be something like "feat: Add support for Gumbel..." then the tooling I use for the changelog will pick it up. 

Thanks for the rebase!

> I am not sure of any other Rust statistics lib that does Gumbel distribution. Sorry :(

That's okay, I don't either. Just wanting to include potential extreme values that are included in other implementations (language agnostic) of Gumbel. I didn't find values tested against in the scripy suite, just a comparison to the Kappa4 distribution. 

Alternatively, you can choose some values that we should expect to be extreme to test each of the functions against. Motivating how large "extreme" is would be up to you. Floating point math is mostly magic to me.

**Comment by soumyasen1809** [comment]
Hi @YeungOnion thanks for the reply.

> Yes, if just the first of these commits could instead be something like "feat: Add support for Gumbel..." then the tooling I use for the changelog will pick it up.

Ok, got it... Pushed now [feat: Add support for Gumbel distribution](https://github.com/statrs-dev/statrs/pull/285/commits/dbbec67b72ef21494c28d48f166e01499fca645f)

> Alternatively, you can choose some values that we should expect to be extreme to test each of the functions against. Motivating how large "extreme" is would be up to you. Floating point math is mostly magic to me.

I have to check that. I am not sure if/how I can find it. Since most of the values were matching till at least `1e-12`. I guess this analysis can be done later, after the PR is merged.

**Comment by YeungOnion** [comment]
Agreed, it's hard to know hard far into the tail users will need to evaluate it.

**Comment by YeungOnion** [comment]
Our CI caught a need for conditional compile on the "rand" feature, this was just added in #275 so no worries that you missed it.

[Rust conditional compilations](https://doc.rust-lang.org/reference/conditional-compilation.html) allows for including/excluding code based on some predicates, there's a crate feature that allows `rand` to be a dependency, so when CI ran the job excluding the feature, there was no dependency named `rand` and it failed.

Would you add this `#cfg[...]` to the implementation of the `rand::distributions::Distribution` trait so `statrs` can compile with default features off?

**Comment by soumyasen1809** [comment]
> Would you add this `#cfg[...]` to the implementation of the `rand::distributions::Distribution` trait so `statrs` can compile with default features off?

Ok, so I added
```rust
#[cfg(feature = "rand")]
impl ::rand::distributions::Distribution<f64> for Gumbel {...}
```

**Review by YeungOnion — APPROVED** [review]

**Comment by YeungOnion** [comment]
> We now have another distribution,
> It looks good.
> I thank you for your contribution!

File: src/distribution/gumbel.rs [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+    /// # Formula
+    ///
+    /// ```text
+    /// 1 - e^(-e^(-(x - μ) / β))
+    /// ```
+    ///
+    /// where `μ` is the location and `β` is the scale
+    fn sf(&self, x: f64) -> f64 {
+        1.0 - (-(-(x - self.location) / self.scale).exp()).exp()
+    }
```
[the remark this answers is not in the document]
Comment (soumyasen1809) [reply src/distribution/gumbel.rs:171]: @YeungOnion Hi, I fixed it in my next PR
https://github.com/statrs-dev/statrs/pull/294

**Pull request merged**